### PR TITLE
infra: set English locale for tests

### PR DESCRIPTION
**Pull request by ZuevKirill95**
related to https://github.com/checkstyle/checkstyle/pull/11529

### CODE REVIEW

**Comment by ZuevKirill95** [comment]
Error:
https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=10743&view=logs&j=69ff11bc-5a77-53a4-545f-8dfc2c866a0e&t=fe4ed25a-cc1d-5fd5-85b9-28f4310f8666&l=20

```
HEAD is now at 12c208331 infra: run checkstyle in ru locale
The PR is based on a master that is 25 commit(s) old.
The max allowed is 10.
This PR is too old and should be rebased on origin/master.
```

I suppose it because commit infra: run checkstyle in ru locale too old
Maybe remove this commit? Same changes in https://github.com/checkstyle/checkstyle/pull/12039

**Comment by rnveach** [comment]
@ZuevKirill95 You have to rebase on the latest master and remove all merge commits.

> Maybe remove this commit? Same changes in https://github.com/checkstyle/checkstyle/pull/12039

I am not understanding why we have the same commit in 2 PRs. If one is just for testing then please point it out, but before removing any commits this PR needs confirmation the changes will work before merging.

**Comment by ZuevKirill95** [comment]
> I am not understanding why we have the same commit in 2 PRs.

I'm not sure why
@romani asked me for this in https://github.com/checkstyle/checkstyle/pull/11529#issuecomment-1211483451
Maybe I misunderstood

**Comment by rnveach** [comment]
I thought my understanding about 11529 was that it was to be closed in favor of this PR as we were using that as a test area to identify the issue.

He event said in the comment:
> As you open your PR, I will close my PR.

So yes, 11529 is going away.

**Comment by romani** [comment]
We need to preserve changes at https://github.com/checkstyle/checkstyle/pull/11529/commits they show correct way to proporate locale, please squash them in one commit 

Second commit should be from Kirill

**Comment by ZuevKirill95** [comment]
Help with this error, please
image

**Comment by ZuevKirill95** [comment]
I added "Dsurefire" to witelist

**Comment by ZuevKirill95** [comment]
It seems this error is not related to my changes

image

**Comment by romani** [comment]
@Vyom-Yadav , @nick-mancuso  is this known instability, can we register it in suppression xml  ? to not bother us unexpectedly

**Comment by romani** [comment]
I placed one more commit with Albanian language - https://www.oracle.com/java/technologies/javase/jdk8-jre8-suported-locales.html
we unlikely to get translation to this language :).

**Review by romani — APPROVED** [review]
ok to merge

**Comment by nick-mancuso** [comment]
> @Vyom-Yadav , @nick-mancuso is this known instability, can we register it in suppression xml ? to not bother us unexpectedly

I have not seen this particular failure previously, I have noticed a couple of pitest-ant failures and one pitest-javadoc failure lately: https://github.com/checkstyle/checkstyle/actions/workflows/pitest.yml?query=branch%3Amaster+is%3Afailure

Failures have been random enough that I do not think we need to consider "unstable" option yet, I do not want to overuse/ abuse this attribute.

**Comment by romani** [comment]
Does it make sense to collect all instabilities? to notice that they likely to happen, and may be not that often.
Do you have already some issue for this ?

**Comment by nick-mancuso** [comment]
> Does it make sense to collect all instabilities? to notice that they likely to happen, and may be not that often. Do you have already some issue for this ?

We have one for pitest-coding only, but we are preparing to close this. I will create new general tracker.

**Comment by romani** [comment]
```
Pitest-api


New surviving mutation(s) found:

Source File: "Violation.java"
Class: "com.puppycrawl.tools.checkstyle.api.Violation$Utf8Control"
Method: "newBundle"
Line Contents: "final String resourceName = toResourceName(bundleName, &quot;properties&quot;);"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.experimental.ArgumentPropagationMutator"
Description: "replaced call to com/puppycrawl/tools/checkstyle/api/Violation$Utf8Control::toResourceName with argument"
Line Number: 542
Error: Process completed with exit code 1.
```

Restarted

**Comment by nick-mancuso** [comment]
https://github.com/checkstyle/checkstyle/issues/12072

**Comment by romani** [comment]
Strange , stable failure.
```
Source File: "Violation.java"
Class: "com.puppycrawl.tools.checkstyle.api.Violation$Utf8Control"
Method: "newBundle"
Line Contents: "final String resourceName = toResourceName(bundleName, &quot;properties&quot;);"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.experimental.ArgumentPropagationMutator"
Description: "replaced call to com/puppycrawl/tools/checkstyle/api/Violation$Utf8Control::toResourceName with argument"
Line Number: 542
```

This looks like related to code changes of this PR.
@ZuevKirill95 , can you look what we can do do kill survival?
Might be extra test is required with localized execution.

**Comment by ZuevKirill95** [comment]
I'm new to pit tests

I don't understand how to reproduce this on the local computer
I tried to change this code 

```java
 final String resourceName = toResourceName(bundleName, "properties");
```

to

```java
 final String resourceName = bundleName;
```
and to
```java
 final String resourceName = "properties";
```

but the tests failed. Does this mean that the mutation should not have survived?

**Comment by Vyom-Yadav** [comment]
@ZuevKirill95 Strange, you have got the mutation correct, I cross-checked it, and the actual mutation is `final String resourceName = "properties";` but the test suite is failing with that, however, pitest shows it as survival.

**Comment by romani** [comment]
@Vyom-Yadav , Should we add this survival to suppression file ?

**Comment by Vyom-Yadav** [comment]
@ZuevKirill95 Strangely using `@BeforeEach` instead of `@BeforeAll` kills the mutation, the issue for the same has been opened at https://github.com/hcoles/pitest/issues/1081.

For now, you can use `@BeforeEach` to kill the mutation.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: This is turning into an ugly issue. `ViolationTest` has a static final Locale which gets its value before this method is run. So now we can have a test global variable with its own Locale and then this Before method comes in and set it to its own specific value and possibly conflict with each other. If they both have the same value, there is no issue, but if they are different then this could become messy.

@romani @nick-mancuso In my own project, I had to create a rule that "static final fields" for tests must be a single state throughout the entire life of the tests and can never change. I really found this to be helpful when I introduced random test ordering (https://github.com/checkstyle/checkstyle/issues/4818) as I found bad tests that were modifying values that should have been constant throughout the entire test run and was causing unpredictable issues.  I also forbid mutable fields for tests so 1 test can be completely isolated, but I am willing for this to be ignored for now.

I think we need to incorporate this rule and reconsider these fields, probably starting with this PR.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: https://github.com/checkstyle/checkstyle/blob/fe3969a2f16854a93d66e80e0ac2df3aa5a3102c/src/test/java/com/puppycrawl/tools/checkstyle/api/ViolationTest.java#L55

Is used only 3 times , we can move definition to each test , to let test control locale .

Goal of this PR to let people with non EN locale to run tests/build without problems. Benefits of this is singnificatly bigger than potential problem that we can resolve later on by overrides and any other way.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: All tests which want to play with locale can have separate base class , and do what makes sense to them.

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: > we can move definition to each test 

It will most likely be more involved than this. We are setting the default Locale probably after all our production classes have loaded. Some like Violation, use the preset locale which is loaded before the before test is run. This is starting to look like all tests need to extend a base and we need a basic init routine.

Comment (ZuevKirill95) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: My suggetion add `@DefaultLocale("en")` in each test

This is convenient for those who have a different locale, if you need to locally check that the message is correct in your language, just change locale in test

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: This won't properly handle our production code which is what I was referring to in my previous post.
Example: https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/Violation.java#L68

Just for noting, `@DefaultLocale` saves the current default locale before the test is run, switches the locale to the specified in the annotation, and the restores the previous default locale after the test is run.

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: IMO, I don't think we should do this BeforeEach and setting the locale in it. We will have to setup and maintain more than just the locale for our entire suite to be in that language.

The whole reason for setting the locale to english is because some tests require english language to validate properly.

My suggestion is we modify tests to skip validation when English is required and language in use is not English. The environment is already in place that the language coming in is used everywhere. We fixed the issue with our CI to run tests in any locale. Lets just skip tests that test English specific results.

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: Another reason to not use `Locale.setDefault`. https://github.com/checkstyle/checkstyle/issues/7436

**Comment by rnveach** [comment]
I am marking this as blocked since I don't think its safe to merge in `Locale.setDefault(Locale.ENGLISH);`. We should be thinking about something like #12104 to resolve the underlying issue.

If the rest of the changes can be separated away and won't cause harm, then I am good to merge them in their own separate PR.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: > Lets just skip tests that test English specific results.

as we use english messages in Input files, we are forced to run in EN locale.

**Comment by rnveach** [comment]
@romani I see it as pointless to continue this PR. My prior denial and block message has not changed. Even the fixes I see you doing now don't fix the immutable issue that I mentioned in my post and that is why it continues to fail.

**Comment by romani** [comment]
locale based test are [failing](https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=17975&view=logs&j=f07bbe47-a53d-57aa-396b-b864e923835d&t=067af9d2-0932-5b80-cadd-10bf65a9f15a&l=1321):
```
[ERROR] Failures: 
[ERROR]   LocalizedMessageTest.testNullArgs:60 Violation should contain exception info
value of           : getMessage()
expected to contain: Error auditing myfile
but was            : Fehler beim Prüfen von myfile
```

**Comment by romani** [comment]
@rnveach, I got a stable form of test executions.

this is good step forward before #12104 because:
- resolves problems for non english contributors (my laptop), it might not be very stable but a way better then stable failure of `mvn test`
- fix our CI test config to finally execute test in locale
- all hacks are located in one commit so it in #12104 we simply revert it as first commit and do what is required to do

and I am switching to brain storm to #12104 that might take a bit to agree by all as it is api changes.

**Comment by rnveach** [comment]
Code has not changed, so my review hasn't changed and this should still be blocked. Even admitting we will revert most of this PR shows this isn't a good step forward. I am against adding more hacks and promising fixes later when the later never comes.

**Comment by romani** [comment]
Right now , and it always was like this, we can pass `mvn test` only on English locale.
We do not change anything. We simply making it work. Nothing new, all playing with locale is not working fake, and now we finally make it clear.

Extra commit, makes life on bigger part of contributors easy and it doesn't not change anything for you.

**Comment by romani** [comment]
@nrmancuso , we are in disagreement, please share second opinion on this PR.
My position is to ease problem for all contributors while we dealing with big gap in our design for locales that existed forever.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java
Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: @rnveach 

>Before method comes in and set it to its own specific value and possibly conflict with each other. If they both have the same value, there is no issue, but if they are different then this could become messy.

Are we able to make assertions about this case (different locale)?

>My suggestion is we modify tests to skip validation when English is required and language in use is not English.

can you expand on this?

**Comment by nrmancuso** [comment]
Honestly, I am pretty confused about what's going on here. We are referencing closed/blocked/discussion ongoing issue. Why do we need the change in command line args? Can this be merged separately?

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: > Are we able to make assertions about this case (different locale)?

If you are talking about code assertions, then yea they can always be added, but it doesn't prevent the issue. And then another issue is what if new code comes in. We won't remember to add more of these checks and the issue could become worse, that is why this isn't a good change.

> My suggestion is we modify tests to skip validation when English is required and language in use is not English.
> can you expand on this?

From what I remember of the issue, basically when a non-English user runs our tests there are failures related to language. Some test(s) is expecting English, and the result is actually the user's non-English locale.

I believe my suggestion was to drop all the code in this PR, and skip over the current test failures ONLY if the user's locale isn't English and validate as many other things as we can. This wouldn't skip anything in CI, as CI doesn't fail on this issue currently.
Basically:
````
if (language == English) {
  validate(....);
}
````

My other suggestion is to actually fix the issue, which is what my other PRs (https://github.com/checkstyle/checkstyle/pull/12511 and https://github.com/checkstyle/checkstyle/pull/12560) were doing. In those PRs, since Checker controls the locale currently (https://checkstyle.org/config.html#Checker_Properties), it would pass it to all the other modules for when a locale is needed and we would drop the global static which is the center of the issue. Basically switching a global static, to a stack based.

**Review by nrmancuso — CHANGES_REQUESTED** [review]
I am ok with these hacks to stabilize test execution for non-english users. This issue doesn't personally affect me, so it is hard for me to weigh risks vs benefits. Ease of contribution workflow provides value to users, contributors, and maintainers alike.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: image

it is execution before all test methods of all tests.
We can probably can do some other elegant way, but as we consider this as temporal update, let keep it.
I do not know how to make it better.

File: src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:1839]: done, but the more I think on this, the more I become confident that all test should be english, only rare test methods that does locale specific testing can be different.

File: src/test/java/com/puppycrawl/tools/checkstyle/DefaultLoggerTest.java
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/DefaultLoggerTest.java]: done

**Comment by romani** [comment]
> Also, we need agreement to also take an iterative approach to this issue in general, and move forward with https://github.com/checkstyle/checkstyle/pull/12511 even if it is not perfect.

I already posted comment to it, waiting for reply.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java
Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:54]: Ok, yes, I see now that `AbstractModuleTestSupport` extends this class, I am good.

File: src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java
Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:1839]: >done, but the more I think on this, the more I become confident that all test should be english, only rare test methods that does locale specific testing can be different.

Agreed

**Review by nrmancuso — APPROVED** [review]
I am good, update is not perfect but at least it is a step forward. @romani please make CI happy

**Comment by romani** [comment]
@rnveach , please consider PR one more time.

**Comment by rnveach** [comment]
My point of view that this PR isn't a good change has not changed.

**Comment by romani** [comment]
This very rare case, but I need to use super power to override disagreement, to move energy to fix in better way at next PR, https://github.com/checkstyle/checkstyle/pull/12511 has to be changed a bit to avoid API changes.

@nrmancuso , please comment on that PR too with your thoughts.

**Pull request merged**
